src/query/list_file.c

```diff
@@ -1688,8 +1690,7 @@ qfile_generate_tuple_into_list (THREAD_ENTRY * thread_p, QFILE_LIST_ID * list_id
       return ER_FAILED;
     }
 
-  offset = ((tuple_length + list_id_p->last_offset) > DB_PAGESIZE ? DB_PAGESIZE : tuple_length);
-  qfile_add_tuple_to_list_id (list_id_p, page_p, tuple_length, offset);
+  qfile_add_tuple_to_list_id (list_id_p, page_p, tuple_length, MIN (tuple_length, qfile_Max_tuple_page_size));
```

Comment:
qfile_generate_tuple_into_list doesn't seem really safe. qfile_save_tuple cannot work if tuple_length > qfile_Max_tuple_page_size. To find potential issues we should:

check before qfile_allocate_new_page_if_need: assert (tuple_length <= qfile_Max_tuple_page_size);

check after qfile_save_tuple: assert (page_p + tuple_length - cur_page_p <= DB_PAGESIZE); // tuple_length can increase in T_MERGE case).

qfile_add_tuple_to_list_id can be given tuple_length argument since it will fail anyway if its size is more than qfile_Max_tuple_page_size.